### DISCUSSION TITLE
How to add new classes/structures to OpenROAD's schema?

### DISCUSSION THREAD

**Discussion by fgaray** [The-OpenROAD-Project/OpenROAD#4269] [General]
Hello OpenRoad devs,

I am following https://github.com/The-OpenROAD-Project/OpenROAD/discussions/3619 to implement a structure to share scan chain data between OpenROAD's components but I am blocked at an compile issue.

See: https://github.com/fgaray/OpenROAD/tree/ctl_db, path src/odb/src/codeGenerator/schema/scan

I am adding 3 new *.json files: dbScanInst.json dbScanPartition.json dbScanPin.json

In dbScanPin.json I am defining an "union" of dbBTerm and dbITerm

In dbScanPartition.json I am trying to use dbScanPin for my "start" and "stop" fields.

I am running the ./generate script in src/odb/src/codeGenerator to generate the C++ code but I am getting the following error:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/1619948/fc8774d8-b742-4407-86ef-e877d347f5d3)

Looks like it is trying to use dbScanPin in the private dbScanPartition.h class, but if I change it to _dbScanPin in my  dbScanPartition.json, then it fails in the public class definition in db.h.

Is there an step that I am missing or maybe some config?

Thanks!

**Comment by maliberty** [comment]
Use dbId to store references to other db objects
```
       "name": "start",
-      "type": "dbScanPin"
+      "type": "dbId<dbScanPin>"
     },
     {
       "name": "stop",
-      "type": "dbScanPin"
+      "type": "dbId<dbScanPin>"
     },
```

**Reply by maliberty** [reply]
You'll want to mark them private as well and write custom getter/setter methods
```
    {
      "name": "start_",
      "type": "dbId<dbScanPin>",
      "flags":["private"]
    },
    {
      "name": "stop_",
      "type": "dbId<dbScanPin>",
      "flags":["private"]
    },
```

**Comment by fgaray** [comment]
I forgot to close this issue, but the answer was given by @maliberty .

Writing the custom setter was the way to go when implementing this.
